sockets/lib/multiplexer.go

```diff
@@ -0,0 +1,334 @@
+/**
+ * Multiplexer - Socket/Channel/Subchannel state machine
+ * https://pokemonshowdown.com/
+ *
+ * This keeps track of the sockets that connect to the SockJS server. Sockets
+ * are stored in the multiplexer to allow the parent process to manipulate them
+ * as it pleases. Channels represent rooms in the parent process; subchannels
+ * split battle rooms into three groups: side 1, side 2, and spectators.
+ * Certain messages will display differently depending on which subchannel the
+ * user's socket is in.
+ */
+
+package sockets
+
+import (
+	"fmt"
+	"net"
+	"path"
+	"regexp"
+	"strconv"
+	"sync"
+	"sync/atomic"
+
+	"github.com/igm/sockjs-go/sockjs"
```

Comment:
The stable version doesn't have `*session.Request`,yet, which is needed to get the socket's remote address and `X-Forwarded-For` headers to pass back to the parent process